Index: backports-4.19.7-1/net/wireless/chan.c
===================================================================
--- backports-4.19.7-1.orig/net/wireless/chan.c
+++ backports-4.19.7-1/net/wireless/chan.c
@@ -931,8 +931,13 @@ static bool _cfg80211_reg_can_beacon(str
 				     bool check_no_ir)
 {
 	bool res;
+
+    /*
 	u32 prohibited_flags = IEEE80211_CHAN_DISABLED |
 			       IEEE80211_CHAN_RADAR;
+    */
+
+	u32 prohibited_flags = IEEE80211_CHAN_DISABLED;
 
 	trace_cfg80211_reg_can_beacon(wiphy, chandef, iftype, check_no_ir);
 
