Index: hostapd-2019-08-08-ca8c2bd2/src/ap/dfs.c
===================================================================
--- hostapd-2019-08-08-ca8c2bd2.orig/src/ap/dfs.c
+++ hostapd-2019-08-08-ca8c2bd2/src/ap/dfs.c
@@ -833,7 +833,7 @@ int hostapd_dfs_complete_cac(struct host
 		 * If it brings more questions, I am very sorry and please
 		 * tell me and remove it.
 		 */
-		hostapd_setup_interface_complete(iface, 1);
+		//hostapd_setup_interface_complete(iface, 1);
 	}
 
 	return 0;
